packages/data-api/src/AnalyticsFetchQuery.ts

```diff
@@ -65,8 +76,66 @@ const supportsConfig = (aggregationSwitch, config) => {
   );
 };
 
+const orgUnitMapValidator = yup.lazy((entityMap: unknown) => {
+  const validatedMap = yup.object().required().validateSync(entityMap);
+  const fieldValidator = yup.object().shape({ code: yup.string().required() });
+  return yup.object().shape(
+    Object.keys(validatedMap).reduce<Record<string, typeof fieldValidator>>(
+      (mapValidator, key: string) => ({
+        ...mapValidator,
+        [key]: fieldValidator,
+      }),
+      {},
+    ),
+  );
+});
+
+export const buildEntityMap = (orgUnitMap: unknown = {}) => {
+  const validatedMap = orgUnitMapValidator.validateSync(orgUnitMap);
+
+  return Object.fromEntries(
+    Object.entries(validatedMap).map(([key, value]) => {
+      return [key, value.code];
+    }),
+  );
+};
+
+export type AnalyticsFetchOptions = {
+  dataElementCodes: string[];
+  organisationUnitCodes: string[];
+  startDate?: string;
+  endDate?: string;
+  aggregations: (string | { type?: string; config?: Record<string, unknown> })[] | undefined;
+};
+
+export type Analytic = {
+  period: string;
+  entityCode: string;
+  type: string;
+  value: string;
+  dataElementCode: string;
+};
+
+export type AnalyticsFetchResult = {
+  analytics: Analytic[];
+  numAggregationsProcessed: number;
+};
+
+type QueryAggregation = {
+  config: { entityMap: Record<string, string> } & Record<string, unknown>;
+  switch: AggregationSwitch;
+  stackId: number;
+};
+
 export class AnalyticsFetchQuery {
-  constructor(database, options) {
+  private readonly database: TupaiaDatabase;
```

Comment:
SO GOOD to see private! Yay encapsulation :)

Reply:
I love to using accessors too! It's such a joy when your editor only shows you public methods for auto completion 😻 

Kostas and I chatted about this and we were both quite keen to turn on [this rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-member-accessibility.md) that enforces using them.

I'll add it to the dev-chat agenda since I think it affects the tamanu folks as well, but I'd be happy for us to start enforcing it 👍

Reply:
Yeah that would be great! We probably need to add linting to CI first :)